server/src/termdb.server.init.js

```diff
@@ -455,14 +455,14 @@ export function server_init_db_queries(ds) {
 	}
 
 	q.getSingleSampleData = function (sampleId, term_ids) {
-		const ids = `('${term_ids.join("','")}')`
-
-		const query = `select term_id, value from anno_categorical where sample=? and term_id in ${ids}
-		union all select term_id, value from anno_float where sample=? and term_id in ${ids}
-		union all  select term_id, value from anno_integer where sample=? and term_id in ${ids} `
-
+		let ids = '?,'.repeat(term_ids.length)
+		if (term_ids.length) ids = ids.slice(0, ids.length - 1)
+		else ids = ''
+		const query = `select term_id, value from anno_categorical where sample=? and term_id in (${ids})
+		union all select term_id, value from anno_float where sample=? and term_id in (${ids})
+		union all  select term_id, value from anno_integer where sample=? and term_id in (${ids}) `
```

Comment:
@airenzp What happens if there is no `term_ids.length`, it seems the query will return zero results? There should not be a term_id condition in that case. I think this should be rewritten like this, including specifying a default term_ids argument value of an empty array:

```js
       q.getSingleSampleData = function (sampleId, term_ids=[]) {
	  const termClause = !term_ids.length ? '' : `and term_id in (${term_ids.map(t =>'?').join(','))})`
 		const query = `select term_id, value 
 			from anno_categorical 
 			where sample=? ${termClause}
			union all 
			select term_id, 
			value from anno_float 
			where sample=? ${termClause}
			union all  
			select term_id, value 
			from anno_integer 
			where sample=? ${termClause}`

		const rows = sql.all([sampleId, ...term_ids, sampleId, ...term_ids, sampleId, ...term_ids])
		return rows
	}
```

Reply:
Airen and i talked about this today. i decided rather than running this sync query for sjlife which is scary, should instead read premade text files with data for single samples, as we're doing for mutation>disco. thus no need to modify this query

Reply:
@xzhou82 The anno tables are indexed by sample, there is no table scan, the query is very fast. Here, we even limit the number of terms to be queried, so even faster. No need to complicate with a separate file, that file read would likely be slower than the sqlite read.